trace_decoder/src/iface.rs

```diff
@@ -0,0 +1,185 @@
+//! Public types for this crate.
+//!
+//! These are all in one place because they're about to be heavily refactored.
+
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+
+use ethereum_types::{Address, U256};
+use evm_arithmetization::proof::{BlockHashes, BlockMetadata};
+use keccak_hash::H256;
+use mpt_trie::partial_trie::HashedPartialTrie;
+use serde::{Deserialize, Serialize};
+
+/// Core payload needed to generate proof for a block.
+/// Additional data retrievable from the blockchain node (using standard ETH RPC
+/// API) may be needed for proof generation.
+///
+/// The trie preimages are the hashed partial tries at the
+/// start of the block. A [TxnInfo] contains all the transaction data
+/// necessary to generate an IR.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct BlockTrace {
+    /// The state and storage trie pre-images (i.e. the tries before
+    /// the execution of the current block) in multiple possible formats.
+    pub trie_pre_images: BlockTraceTriePreImages,
+
+    /// A collection of contract code.
+    /// This will be accessed by its hash internally.
+    #[serde(default)]
+    pub code_db: BTreeSet<Vec<u8>>,
+
+    /// Traces and other info per transaction. The index of the transaction
+    /// within the block corresponds to the slot in this vec.
+    pub txn_info: Vec<TxnInfo>,
+}
+
+/// Minimal hashed out tries needed by all txns in the block.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum BlockTraceTriePreImages {
+    /// The trie pre-image with separate state/storage tries.
+    Separate(SeparateTriePreImages),
+    /// The trie pre-image with combined state/storage tries.
+    Combined(CombinedPreImages),
+}
+
+/// State/Storage trie pre-images that are separate.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct SeparateTriePreImages {
+    /// State trie.
+    pub state: SeparateTriePreImage,
+    /// Storage trie.
+    pub storage: SeparateStorageTriesPreImage,
+}
+
+/// A trie pre-image where state & storage are separate.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum SeparateTriePreImage {
+    /// Storage or state trie format that can be processed as is, as it
+    /// corresponds to the internal format.
+    Direct(HashedPartialTrie),
+}
+
+/// A trie pre-image where both state & storage are combined into one payload.
```

Comment:
nit: there's also the representation that differs, the compact Jerigon payloads don't pass in trie structs

Reply:
This is verbatim from the old code, I'll refactor all these structures later